### PR TITLE
Reclaim unused buffer for both traditional and dynamic buffer model

### PR DESCRIPTION
**Pull request by stephenxs**
**What I did**

Reclaim reserved buffer of unused ports for both dynamic and traditional models.
This is done by 
- Removing lossless priority groups on unused ports.
- Applying zero buffer profiles on the buffer objects of unused ports.
- In the dynamic buffer model, the zero profiles are loaded from a JSON file and applied to `APPL_DB` if there are admin down ports.
  The default buffer configuration will be configured on all ports. Buffer manager will apply zero profiles on admin down ports.
- In the static buffer model, the zero profiles are loaded by the buffer template.

Signed-off-by: Stephen Sun <stephens@nvidia.com>

**Why I did it**

**How I verified it**

Regression test and vs test.

**Details if related**
***Static buffer model***

Remove the lossless buffer priority group if the port is admin-down and the buffer profile aligns with the speed and cable length of the port.

***Dynamic buffer model***

****Handle zero buffer pools and profiles****

1. buffermgrd: add a CLI option to load the JSON file for zero profiles.
2. Load them from JSON file into the internal buffer manager's data structure
3. Apply them to APPL_DB once there is at least one admin-down port
   - Record zero profiles' names in the pool object it references.
     By doing so, the zero profile lists can be constructed according to the normal profile list. There should be one profile for each pool on the ingress/egress side.
   - And then apply the zero profiles to the buffer objects of the port.
   - Unload them from APPL_DB once all ports are admin-up since the zero pools and profiles are no longer referenced.
     Remove buffer pool counter id when the zero pool is removed.
4. Now that it's possible that a pool will be removed from the system, the watermark counter of the pool is removed ahead of the pool itself being removed.

****Handle port admin status change****

1. Currently, there is a logic of removing buffer priority groups of admin down ports. This logic will be reused and extended for all buffer objects, including `BUFFER_QUEUE`, `BUFFER_PORT_INGRESS_PROFILE_LIST`, and `BUFFER_PORT_EGRESS_PROFILE_LIST`.
   - When the port is admin down,
     - The normal profiles are removed from the buffer objects of the port
     - The zero profiles, if provided, are applied to the port
   - When the port is admin up,
     - The zero profiles, if applied, are removed from the port
     - The normal profiles are applied to the port.
2. Ports orchagent exposes the number of queues and priority groups to STATE_DB.
   Buffer manager can take advantage of these values to apply zero profiles on all the priority groups and queues of the admin-down ports.
   In case it is not necessary to apply zero profiles on all priority groups or queues on a certain platform, `ids_to_reclaim` can be customized in the JSON file.
3. Handle all buffer tables, including `BUFFER_PG`, `BUFFER_QUEUE`, `BUFFER_PORT_INGRESS_PROFILE_LIST` and `BUFFER_PORT_EGRESS_PROFILE_LIST`
   - Originally, only the `BUFFER_PG` table was cached in the dynamic buffer manager.
   - Now, all tables are cached in order to apply zero profiles when a port is admin down and apply normal profiles when it's up.
   - The index of such tables can include a single port or a list of ports, like `BUFFER_PG|Ethernet0|3-4` or `BUFFER_PG|Ethernet0,Ethernet4,Ethernet8|3-4`. Originally, there is a logic to handle such indexes for the `BUFFER_PG` table. Now it is reused and extended to handle all the tables.
4. [Mellanox] Plugin to calculate buffer pool size:
   - Originally, buffer for the queue, buffer profile list, etc. were not reclaimed for admin-down ports so they are reserved for all ports.
   - Now, they are reserved for admin-up ports only.

****Accelerate the progress of applying buffer tables to APPL_DB****

This is an optimization on top of reclaiming buffer.

1. Don't apply buffer profiles, buffer objects to `APPL_DB` before buffer pools are applied when the system is starting.
   This is to apply the items in an order from referenced items to referencing items and try to avoid buffer orchagent retrying due to referenced table items.
   However, it is still possible that the referencing items are handled before referenced items. In that case, there should not be any error message.
2. [Mellanox] Plugin to calculate buffer pool size:
   Return the buffer pool sizes value currently in APPL_DB if the pool sizes are not able to be calculated due to lacking some information. This typically happens at the system start.
   This is to accelerate the progress of pushing tables to APPL_DB.

### CODE REVIEW

File: cfgmgr/buffermgr.cpp [hunk 1]
```diff
@@ -141,19 +150,52 @@ task_process_status BufferMgr::doSpeedUpdateTask(string port)
...
         return task_process_status::task_invalid_entry;
     }
 
-    // Crete record in BUFFER_PROFILE table
-    // key format is pg_lossless_<speed>_<cable>_profile
-    string buffer_profile_key = "pg_lossless_" + speed + "_" + cable + "_profile";
-
     // check if profile already exists - if yes - skip creation
     m_cfgBufferProfileTable.get(buffer_profile_key, fvVector);
+    // Crete record in BUFFER_PROFILE table
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply cfgmgr/buffermgr.cpp:198]: it should be here but need to fix typo.

File: cfgmgr/buffermgrdyn.cpp [hunk 2]
```diff
@@ -2227,7 +3135,115 @@ void BufferMgrDynamic::doTask(Consumer &consumer)
...
+            {
+                updateBufferProfileToDb(profile.first, profile.second);
+            }
+
+            for (auto &port : m_portInfoLookup)
+            {
+                if (port.second.state != PORT_ADMIN_DOWN)
+                {
+                    refreshPgsForPort(port.first, port.second.effective_speed, port.second.cable_length, port.second.mtu);
+                    for (auto &pg : m_portPgLookup[port.first])
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:3192]: looks like this is not necessary. lossy PGs have been inserted to appl db by refreshPgsForPort

File: cfgmgr/buffermgrdyn.cpp [hunk 3]
```diff
@@ -2227,7 +3135,115 @@ void BufferMgrDynamic::doTask(Consumer &consumer)
...
+        {
+            // Apply all pending buffer profiles
+            for (auto &profile : m_bufferProfileLookup)
+            {
+                updateBufferProfileToDb(profile.first, profile.second);
+            }
+
+            for (auto &port : m_portInfoLookup)
+            {
+                if (port.second.state != PORT_ADMIN_DOWN)
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:3189]: Is it necessary?

Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp:3287]: Yes. Comments added

File: cfgmgr/buffermgrdyn.cpp (hunk 2 above)
Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp]: Removed.

File: cfgmgr/buffermgrdyn.cpp [hunk 4]
```diff
@@ -614,55 +844,98 @@ void BufferMgrDynamic::updateBufferPoolToDb(const string &name, const buffer_poo
...
 
     m_applBufferProfileTable.set(name, fvVector);
     m_stateBufferProfileTable.set(name, fvVector);
 }
 
 // Database operation
-// Set/remove BUFFER_PG table entry
-void BufferMgrDynamic::updateBufferPgToDb(const string &key, const string &profile, bool add)
+// Set/remove BUFFER_PG or BUFFER_QUEUE table entry
+void BufferMgrDynamic::updateBufferObjectToDb(const string &key, const string &profile, bool add, bool isPg=true, bool isReference=false)
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp]: All are references now. No need to add comments

File: cfgmgr/buffermgrdyn.cpp [hunk 5]
```diff
@@ -829,30 +1109,359 @@ bool BufferMgrDynamic::isHeadroomResourceValid(const string &port, const buffer_
...
+        auto const &losslessProfile = portPg.running_profile_name.empty() ? "ingress_lossless_profile" : portPg.running_profile_name;
+        auto const &profileRef = m_bufferProfileLookup.find(losslessProfile);
+        string zeroProfile;
+
+        if (profileRef != m_bufferProfileLookup.end())
+        {
+            auto &profileInfo = profileRef->second;
+            auto &poolInfo = m_bufferPoolLookup[profileInfo.pool_name];
+            zeroProfile = poolInfo.zero_profile_name;
+            profileInfo.port_pgs.erase(key);
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1407]: Should not erase key in case the running_profile_name is empty

File: cfgmgr/buffermgrdyn.cpp [hunk 6]
```diff
@@ -829,30 +1109,359 @@ bool BufferMgrDynamic::isHeadroomResourceValid(const string &port, const buffer_
...
+
+        SWSS_LOG_INFO("Reclaiming buffer reserved for PG %s", key.c_str());
+
+        if (!applyZeroProfileOnSpecifiedPgs)
+        {
+            // Apply zero profiles on each configured PG and supported-but-not-configured PG
+            // Fetch the zero profile of the pool referenced by the configured profile
+            if (!zeroProfile.empty())
+            {
+                SWSS_LOG_INFO("Applying zero profile %s on PG %s of port %s", zeroProfile.c_str(), key.c_str(), port.c_str());
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1424]: ` of port %s` can be removed.

File: cfgmgr/buffermgrdyn.cpp [hunk 7]
```diff
@@ -1919,15 +2711,82 @@ task_process_status BufferMgrDynamic::handleBufferProfileTable(KeyOpFieldsValues
...
+            // It will be updated when its profile configured.
+            zeroProfile = isPg ? m_ingressPgZeroProfileName : m_egressQueueZeroProfileName;
+        }
+        else
+        {
+            auto &profileInfo = searchRef->second;
+            auto &poolInfo = m_bufferPoolLookup[profileInfo.pool_name];
+            zeroProfile = poolInfo.zero_profile_name;
+        }
+        updateBufferObjectToDb(key, zeroProfile, true, false);
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:2756]: `updateBufferObjectToDb(key, zeroProfile, true, isPG);`?

File: cfgmgr/buffermgrdyn.cpp [hunk 8]
```diff
@@ -2052,7 +2896,11 @@ task_process_status BufferMgrDynamic::handleOneBufferPgEntry(const string &key,
             m_bufferProfileLookup[configProfileName].port_pgs.erase(key);
         }
 
-        if (bufferPg.lossless)
+        if (portInfo.state == PORT_ADMIN_DOWN)
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:2899]: check whether removing is supported by the platform

**Review by keboliu — APPROVED** [review]

**Pull request closed without merging**